include/glow/Runtime/HostManager/HostManager.h

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_RUNTIME_HOSTMANAGERR_HOSTMANAGER_H
+#define GLOW_RUNTIME_HOSTMANAGERR_HOSTMANAGER_H
+#include "glow/Backends/Backend.h"
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Runtime/RuntimeTypes.h"
+
+#include <atomic>
+#include <map>
+#include <mutex>
+#include <unordered_map>
+#include <vector>
+
+namespace glow {
+namespace runtime {
+
+/// Callback signalling the result of running a function.
+struct DeviceConfig {
+  std::string deviceName;
+  BackendKind backendKind;
+};
+
+class Executor;
+
+class Provisioner;
+
+/// The HostManager serves as an entry point into the Runtime environment. It
+/// provides an interface to add, run, and evict networks from the host. It
+/// handles DeviceManager initialization, houses the Executor, and calls into
+/// the Partitioner and Provisioner for network initialization.
+class HostManager final {
+  /// Count of current in-flight networks being run. Atomic to allow
+  /// concurrency in runNetwork.
+  std::atomic<size_t> activeRequestCount_{0};
+
+  /// Count of total requests, this is used as a run ID. Atomic to allow
+  /// concurrency in runNetwork.
+  std::atomic<size_t> totalRequestCount_{0};
+
+  /// Limit maximum count of networks run at once. Hardcoded for now this
+  /// should be a configurable value. Above this limite the HostManager will
+  /// refuse additional request and return Failed.
+  const unsigned int activeRequestLimit_ = 20;
+
+  /// A map from a networkName to the DAG that represents the root of a network.
+  std::unordered_map<std::string, std::unique_ptr<DAGNode>> roots_;
```

Comment:
Can we just store map of DAGNodeLists instead of this and networks_?

Reply:
They can have name collision. That was the original design though....

Reply:
I misunderstood, I think that's a great idea, I think it'll be better to land and do the refactor afterwards since to do it right we'd want to move DAGNodeList into runtimeTypes. I've created #2353 to track it.